x-pack/metricbeat/module/azure/client_utils.go

```diff
@@ -222,3 +224,106 @@ func getVM(vmName string, vms []VmResource) (VmResource, bool) {
+	}
+	return resourceIDs
+}
+
+// batchMetricIsEmpty will check if the metric value is empty, this seems to be an issue with the azure sdk
+func batchMetricIsEmpty(metric azmetrics.MetricValue) bool {
+	if metric.Average == nil && metric.Total == nil && metric.Minimum == nil && metric.Maximum == nil && metric.Count == nil {
+		return true
+	}
+	return false
+}
+
+func mapBatchMetricValues(client *BatchClient, metricValues azmetrics.MetricData) []MetricValue {
+	var currentMetrics []MetricValue
+	// compare with the previously returned values and filter out any double records
+	for _, v := range metricValues.Values {
+		for _, t := range v.TimeSeries {
+			for _, mv := range t.Data {
+				if batchMetricIsEmpty(mv) {
+					continue
+				}
+				//// remove metric values that are not part of the timeline selected
+				//if mv.TimeStamp.After(startTime) && mv.TimeStamp.Before(endTime) {
+				//	continue
+				//}
+				// define the new metric value and match aggregations values
+				var val MetricValue
+				val.name = *v.Name.Value
+				val.timestamp = *mv.TimeStamp
+				if mv.Minimum != nil {
+					val.min = mv.Minimum
+				}
+				if mv.Maximum != nil {
+					val.max = mv.Maximum
+				}
+				if mv.Average != nil {
+					val.avg = mv.Average
+				}
+				if mv.Total != nil {
+					val.total = mv.Total
+				}
+				if mv.Count != nil {
+					val.count = mv.Count
+				}
+				if t.MetadataValues != nil {
+					for _, dim := range t.MetadataValues {
+						val.dimensions = append(val.dimensions, Dimension{Name: *dim.Name.Value, Value: *dim.Value})
+					}
+				}
+				currentMetrics = append(currentMetrics, val)
+			}
+		}
+	}
+
+	return currentMetrics
+}
+
+// processStore collects and return the metric values of the store using the batchAPI. After the metric values are collected, the store gets cleared.
+func processStore(client *BatchClient, criteria ResDefGroupingCriteria, store *MetricStore, referenceTime time.Time, report mb.ReporterV2) []Metric {
+	groupedMetrics := map[ResDefGroupingCriteria][]Metric{
+		criteria: store.GetMetrics(),
+	}
+	metricValues := client.GetMetricsInBatch(groupedMetrics, referenceTime, report)
+	// client.Log.Infof("metricValues received at %s for criteria %+v", referenceTime, criteria)
+	store.ClearMetrics()
+	return metricValues
+}
+
+// processAllStores collects and return the metrics of all the stores using the batchAPI
+func processAllStores(client *BatchClient, stores map[ResDefGroupingCriteria]*MetricStore, referenceTime time.Time, report mb.ReporterV2) []Metric {
+	var metricValues []Metric
+	for criteria, store := range stores {
+		if store.Size() > 0 {
+			metricValues = append(metricValues, processStore(client, criteria, store, referenceTime, report)...)
+		}
+	}
+	return metricValues
+}
+
+// Function to check if a string is in an array of strings
+func containsString(arr []string, str string) bool {
+	for _, v := range arr {
+		if v == str {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
Can standard library's https://pkg.go.dev/slices#Contains be an alternative?

Reply:
I had in mind to check if there exists such a method and then I forgot. Thanks

Reply:
It was only added in Go 1.21, so it's pretty recent.